### DISCUSSION TITLE
How to speed up polycrystal crystal plasticity simualtion in 3D

### DISCUSSION THREAD

**Discussion by RWTHLHK** [idaholab/moose#29014] [Q&A Modules: Solid mechanics] [answered]
Hello experts. I'd like to perform crystal plasticity simulation in 3D, but I find the polycrystal simulation is quite slow. Is there any suggestion to speed up the process.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Do you have an example simulation? You could profile it and see where the slow path is
https://mooseframework.inl.gov/moose/application_development/profiling.html

@dewenyushu  @jiangwen84 do you have advice on this? Were there any future works performance-wise that could be addressed?

**Reply by RWTHLHK** [reply]
Thanks! I just find the reason is maybe my RVE is kind of too big and contains too many grains.  And I can simply reduce the size of my RVE to a smaller size. My final goal is cross-scale simulation of fracture mechanics, MOOSE is really a convenient tool to achieve this.

**Reply by RWTHLHK** [reply]
I just found it is also possible to wrap external non-MOOSE codes; this can significantly accelerate my progress. But the relevant tutorial is still not in detail, are there any further instructions?

**Reply by GiudGiud** [reply]
there are examples. Cardinal is the biggest one
https://github.com/neams-th-coe/cardinal

**Reply by RWTHLHK** [reply]
OK, thanks. I will learn how Cardinal works.  I want to implement my phase field model coupled with crystal plasticity.  I have used MATLAB's symbolic operations to derive the weak form and Jacobian matrix.  However, the interfacial term of my phase field model is too complex,  it is quite challenging and timeconsuming to code weak form and Jacobian in MOOSE by hand.  Can I externally compute the residual and Jacobian and then pass values to MOOSE?

**Reply by GiudGiud** [reply]
No that's likely going to be too inefficient.
I would recommend using AD if you think the Jacobian takes too long to implement
